tst.c

```diff
@@ -73,98 +51,79 @@ static void *tst_stack_pop(tst_stack *s)
  *  if refcnt non-zero.
  */
-static void *tst_del_word(tst_node **root,
-                          tst_node *node,
-                          tst_stack *stk,
-                          const int freeword)
+void *tst_del_word(tst_stack *stk, const int freeword)
 {
-    tst_node *victim = node;               /* begin deletion w/victim */
-    tst_node *parent = tst_stack_pop(stk); /* parent to victim */
-
-    if (!victim->refcnt) {            /* if last occurrence */
-        if (!victim->key && freeword) /* check key is nul   */
-            free(victim->eqkid);      /* free string (data) */
+    tst_node **pvictim = tst_stack_pop(stk);
+    tst_node *victim = *pvictim;
 
-        /* remove unique suffix chain - parent & victim nodes
-         * have no children. simple remove until the first parent
-         * found with children.
+    if (victim->refcnt > 0) {
+        /* The string is still be referenced.
+         * Return the address of victim.
          */
-        while (!parent->lokid && !parent->hikid && !victim->lokid &&
-               !victim->hikid) {
-            parent->eqkid = NULL;
-            free(victim);
-            victim = parent;
-            parent = tst_stack_pop(stk);
-            if (!parent) { /* last word & root node */
-                free(victim);
-                return (void *) (*root = NULL);
-            }
+        printf("  %s  (refcnt: %u) not removed.\n", (char *) victim->eqkid,
+               victim->refcnt);
+        return victim;
+    }
+
+    /* Free the string in CPY mode.
+     */
+    if (freeword) {
+        free(victim->eqkid);
+    }
+    victim->eqkid = NULL;
+
+    /* remove unique suffix chain - victim
+     * have no children. simple remove until the first node found
+     * with children.
+     */
+    while (!victim->lokid && !victim->hikid && !victim->eqkid) {
+        free(victim);
+        *pvictim = NULL;
+        pvictim = tst_stack_pop(stk);
+        if (!pvictim) {
+            // Stack empty, which means reached the root of the tree.
+            return NULL;
         }
+        victim = *pvictim;
+    }
 
-        /* check if victim is prefix for others (victim has lo/hi node).
-         * if both lo & hi children, check if lokid->hikid present, if not,
-         * move hikid to lokid->hikid, replace node with lokid and free node.
-         * if lokid->hikid present, check hikid->lokid. If not present, then
-         * move lokid to hikid->lokid, replace node with hikid free node.
+    /* Try to rotate the subtree
+     * If 'victim->qekid' is NULL. The 'victim' is a perfix node
+     * of deleting string. In this case, try to rotate the tree.
+     * Otherwise 'victim' represent perfix for another perfix if
+     * 'victim->qekid' isn't NULL. In this case, the delete process
+     * is done and return NULL;
+     */
+    if (victim->eqkid) {  // This node represent another perfix.
+        return NULL;
+    } else if (victim->lokid && victim->hikid) {
```

Comment:
Change `} else if (...) {` to `if (...) {` since the last statement was `return`.